keras_nlp/tokenizers/byte_pair_tokenizer.py

```diff
@@ -0,0 +1,460 @@
+            self._byte_pair_encoding(unseen_tokens)
+            return self.cache.lookup(flatten_tokens)
+
+        # If `count_nonzero_mask == 0`, it means all tokens are in cache, we
+        # just return the lookup result. Otherwise process the unseen tokens.
+        result = tf.cond(
+            count_nonzero_mask == 0, lambda: cache_lookup, process_unseen_tokens
+        )
+
+        # Encode merged tokens.
+        result = tf.strings.split(result, sep=" ")
+        encoding = self.token_to_id_map.lookup(result)
+
+        # Unflatten to match input.
+        encoding = tf.RaggedTensor.from_row_splits(
+            encoding.flat_values,
+            tf.gather(encoding.row_splits, token_row_splits),
+        )
+
+        # Convert to a dense output if `sequence_length` is set.
+        if self.sequence_length:
+            output_shape = encoding.shape.as_list()
+            output_shape[-1] = self.sequence_length
+            encoding = encoding.to_tensor(shape=output_shape)
+        # Convert to a dense output if input in scalar
+        if scalar_input:
+            encoding = tf.squeeze(encoding, 0)
+            tf.ensure_shape(encoding, shape=[self.sequence_length])
+
+        return encoding
+
+    def detokenize(self, inputs):
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+
+        scalar_input = inputs.shape.rank == 0
+        if scalar_input:
+            inputs = tf.expand_dims(inputs, 0)
+
+        unicode_text = tf.strings.reduce_join(
+            self.id_to_token_map.lookup(inputs), axis=1
+        )
+        split_unicode_text = tf.strings.unicode_split(unicode_text, "UTF-8")
+        byte_text = tf.strings.reduce_join(
+            self.unicode2byte.lookup(split_unicode_text)
+        )
+
+        if not scalar_input:
+            byte_text = tf.expand_dims(byte_text, 0)
+
+        return byte_text
+
+    def _encode_tokens(self, tokens):
+        """Map token bytes to unicode using `byte2unicode`."""
+        split_bytes = tf.strings.bytes_split(tokens)
+        split_unicode = self.byte2unicode.lookup(split_bytes)
+        return split_unicode
+
+    def _remove_strings(self, tensor, string_to_remove):
+        """Remove certain strings from input tensor."""
+        non_empty_mask = tensor != string_to_remove
+        flatten_indexes = tf.where(non_empty_mask)
+        flatten_result = tf.gather_nd(tensor, flatten_indexes)
+        row_lengths = tf.reduce_sum(tf.cast(non_empty_mask, tf.int64), axis=1)
+        result = tf.RaggedTensor.from_row_lengths(
+            values=flatten_result,
+            row_lengths=row_lengths,
+        )
+        return result
+
+    def _remove_empty_strings(self, tensor):
+        """Remove empty strings in a tensor"""
+        return self._remove_strings(tensor, "")
+
+    def _remove_whitespace_placeholder(self, tensor):
+        return self._remove_strings(tensor, "६")
+
+    @tf.function
```

Comment:
why do we need a tf.function here? won't `tf.data` compile this normally?

Reply:
It's required, deleting this causes an error, which I think is related to the implementation.